pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,383 @@
+	if err := checkForCompleteBlock(ctx, logger, bULID, userBkt); err != nil {
+		checkBlockUploadError(err, op, "while checking for complete block", logger, w)
+		return
+	}
+
+	if shouldComplete {
+		err = c.completeBlockUpload(ctx, r, logger, userBkt, tenantID, bULID)
+	} else {
+		err = c.createBlockUpload(ctx, r, logger, userBkt, tenantID, bULID)
+	}
+	if err != nil {
+		checkBlockUploadError(err, op, "", logger, w)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+func checkBlockUploadError(err error, op, extra string, logger log.Logger, w http.ResponseWriter) {
+	var httpErr httpError
+	if errors.As(err, &httpErr) {
+		level.Warn(logger).Log("msg", httpErr.message, "operation", op)
+		http.Error(w, httpErr.message, httpErr.statusCode)
+		return
+	}
+
+	if extra != "" {
+		extra += " "
+	}
+	level.Error(logger).Log("msg", fmt.Sprintf("an unexpected error occurred%s", extra), "operation", op,
+		"err", err)
+	http.Error(w, "internal server error", http.StatusInternalServerError)
+}
+
+// checkForCompleteBlock checks for a complete block with same ID. If one exists, an error is returned.
+func checkForCompleteBlock(ctx context.Context, logger log.Logger, blockID ulid.ULID, userBkt objstore.Bucket) error {
+	exists, err := userBkt.Exists(ctx, path.Join(blockID.String(), block.MetaFilename))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to check existence of %s in object storage", block.MetaFilename))
+	}
+	if exists {
+		return httpError{
+			message:    "block already exists in object storage",
+			statusCode: http.StatusConflict,
+		}
+	}
+
+	return nil
+}
+
+func (c *MultitenantCompactor) createBlockUpload(ctx context.Context, r *http.Request,
+	logger log.Logger, userBkt objstore.Bucket, tenantID string, blockID ulid.ULID) error {
+	level.Debug(logger).Log("msg", "starting block upload")
+
+	meta, err := decodeMeta(r.Body, "request body")
+	if err != nil {
+		return httpError{
+			message:    "malformed request body",
+			statusCode: http.StatusBadRequest,
+		}
+	}
+
+	if msg := c.sanitizeMeta(logger, tenantID, blockID, &meta); msg != "" {
+		return httpError{
+			message:    msg,
+			statusCode: http.StatusBadRequest,
+		}
+	}
+
+	return c.uploadMeta(ctx, logger, meta, blockID, tmpMetaFilename, userBkt)
+}
+
+// UploadBlockFile handles requests for uploading block files.
+//
+// It takes the mandatory query parameter "path", specifying the file's destination path.
+func (c *MultitenantCompactor) UploadBlockFile(w http.ResponseWriter, r *http.Request) {
+	const op = "block file upload"
+
+	vars := mux.Vars(r)
+	blockID := vars["block"]
+	bULID, err := ulid.Parse(blockID)
+	if err != nil {
+		http.Error(w, "invalid block ID", http.StatusBadRequest)
+		return
+	}
+	pth := r.URL.Query().Get("path")
+	if pth == "" {
+		http.Error(w, "missing or invalid file path", http.StatusBadRequest)
+		return
+	}
+
+	tenantID, ctx, err := tenant.ExtractTenantIDFromHTTPRequest(r)
```

Comment:
Let's use `tenant.TenantID(r.Context())` instead, to be consistent with other parts of Mimir.

Reply:
Done.